### DISCUSSION TITLE
Self-built containers may be needed to reduce memory cost

### DISCUSSION THREAD

**Discussion by kirk0830** [deepmodeling/abacus-develop#3076] [General]
### Background
1. Many data are stored in HTML-formatted file, for example the coefficients of spherical bessel function [bessel_basis.cpp:line 337](https://github.com/deepmodeling/abacus-develop/blob/develop/source/module_io/bessel_basis.cpp#LL337C24-L337C24). To access data with higher efficiency, rather than search for data in certain line with time complexity O(N^2), HTML parser with mapping strategy will be a good choice.
2. In input module, the way to parse INPUT file now is still the one-by-one strcmp method. Map can also be used here to reach a at least O(N^2) to O(NlogN) speed-up.

### Problem
the C++ STL has non-negligible extra memory cost (including std::vector, std::map, etc), this will bring about unexpected memory cost problem to the overall program. To reproduce, use the toy HTML parser I write to parse UPF-formatted pseudopotential file. Although in principle the speed of accessing data will be higher, but memory cost now is terribly high.
[html_parser.zip](https://github.com/deepmodeling/abacus-develop/files/11608094/html_parser.zip)

### Possible solution
Build containers by ourselves, so as to avoid extra memory cost by redundant functions and data.

**Comment by kirk0830** [comment]
💡, I will let the constructor of HTML class require a set of names of tags whose contents will be saved in memory, for all other tags there contents will be directly omitted.
But the self-built containers will still be important.

**Comment by hongriTianqi** [comment]
- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

**Comment by hongriTianqi** [comment]
This will be needed in generating orbitals. It will be solved later.

**Comment by hongriTianqi** [comment]
@jinzx10
